java/ql/src/experimental/semmle/code/java/security/StaticInitializationVectorQuery.qll

```diff
@@ -0,0 +1,166 @@
+import java
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.dataflow.TaintTracking2
+
+/**
+ * Holds if `array` is initialized only with constants, for example,
+ * `new byte[8]` or `new byte[] { 1, 2, 3, 4, 5, 6, 7, 8 }`.
+ */
+private predicate initializedWithConstants(ArrayCreationExpr array) {
+  not exists(array.getInit())
+  or
+  forex(Expr element | element = array.getInit().getAChildExpr() |
+    element instanceof CompileTimeConstantExpr
+  )
+}
+
+/**
+ * An expression that creates a byte array that is initialized with constants.
+ */
+private class StaticByteArrayCreation extends ArrayCreationExpr {
+  StaticByteArrayCreation() {
+    this.getType().(Array).getElementType().(PrimitiveType).getName() = "byte" and
```

Comment:
Might be better to use [`getComponentType()`](https://codeql.github.com/codeql-standard-libraries/java/semmle/code/java/Type.qll/predicate.Type$Array$getComponentType.0.html) instead of `getElementType()`, because the above would also hold for `byte[][][]` while with `getComponentType()` it would only hold for `byte[]`.

Reply:
This expression doesn't check the dimension. Both methods work fine here, I don't really see why `getComponentType()` might be better here. The docs for `getElementType()` [says](https://codeql.github.com/codeql-standard-libraries/java/semmle/code/java/Type.qll/predicate.Type$Array$getElementType.0.html):

> Gets the type of the elements used to construct this array type.
> 
> For example, the element type of Object[][] is Object.

That's exactly what is necessary here.

Reply:
But will a multi-dimensional `byte` array (e.g. `byte[][][]`), or one of its sub arrays, ever be used as initialization vector (and would the other parts of this query even handle that correctly)?

Because `StaticByteArrayCreation` is currently matching that, e.g. `new byte[5][]`.

Reply:
I think Marcono is right, you really do only want `byte[]` here

Reply:
I didn't consider multi-dimensional arrays. I think it may make sense to support them. For example, an application can use a set of static IVs like this:

```java
byte[][] ivs = new byte[][] {
  new byte[] { 1, 2, 3 ... },
  new byte[] { 4, 5, 6, ...}
  ...
}
...
GCMParameterSpec ivSpec = new GCMParameterSpec(128, ivs[i]);
```

I added test cases `encryptWithOneOfStaticIvs01`, `encryptWithOneOfStaticIvs02` and `encryptWithOneOfStaticZeroIvs` that use multi-dimensional arrays, and updated the `initializedWithConstants()` predicate.

`array.getInit().getAnInit().getAChildExpr() instanceof IntegerLiteral` covers the following case:

```java
byte[][] ivs = new byte[][] {
  new byte[8],
  new byte[16]
};
```

This expression looks weird to me, but I could not find a better way to cover the case above. While debugging this case, I saw the child elements of the `ArrayInit`s are instances of `IntegerLiteral` and `TypeAccess`. Please let me know if there is a better way.

`StaticByteArrayCreation` still uses `getElementType()` because it wants to check only the type but not dimensions. If we decide to cover only one-dimension arrays, then we can use `getElementType()`. I would prefer to support multi-dimensional arrays though.

Reply:
>While debugging this case, I saw the child elements of the ArrayInits are instances of IntegerLiteral and TypeAccess.

The VSCode extension AST viewer shows that for me as well, and manually quering `ArrayCreationExpr` and `ArrayInit` seems to confirm that.
To me that looks like a bug in the extractor, that `ArrayCreationExpr` inside an `ArrayInit` is modelled as (malformed) `ArrayInit`. Might be worth creating an issue for this?

Reply:
> To me that looks like a bug in the extractor, that ArrayCreationExpr inside an ArrayInit is modelled as (malformed) ArrayInit. Might be worth creating an issue for this?

Yes, I also expected to see `ArrayCreationExpr` inside. @smowton Does it look like a bug? Or, are we missing something?

Reply:
Looks like a bug

Reply:
Created https://github.com/github/codeql/issues/6552#issue-979422575